File: 2020/day05/ex2.py
import sys

lines = [l.strip() for l in sys.stdin]
found_seats = set()
for line in lines:
    min_row = 0
    max_row = 127
    min_col = 0
    max_col = 7

    for c in line:
        if c == 'F':
            max_row = int(min_row + (max_row - min_row) / 2)
        elif c == 'B':
            min_row = int(min_row + (max_row - min_row + 1) / 2)
        elif c == 'L':
            max_col = int(min_col + (max_col - min_col) / 2)
        elif c == 'R':
            min_col = int(min_col + (max_col - min_col + 1) / 2)

    seat_id = min_row * 8 + min_col
    found_seats.add(seat_id)

for row in range(127):
    for col in range(7):
        seat_id = row * 8 + col
        if (seat_id not in found_seats
            and seat_id - 1 in found_seats
            and seat_id + 1 in found_seats):
            print(seat_id)

File: 2020/day05/ex2.c
#include <stdio.h>
#include <strings.h>

int main() {
  char buffer[64];
  char found_seats[128];
  int i;

  bzero(found_seats, sizeof(found_seats));

  while (fgets(buffer, sizeof(buffer), stdin)) {
    char *c;
    int min_row = 0;
    int max_row = 127;
    int min_col = 0;
    int max_col = 7;

    for (c = buffer; *c && *c != '\n'; c++) {
      switch (*c) {
        case 'F':
            max_row = min_row + ((max_row - min_row) >> 1);
            break;
        case 'B':
            min_row = min_row + ((max_row - min_row + 1) >> 1);
            break;
        case 'L':
            max_col = min_col + ((max_col - min_col) >> 1);
            break;
        case 'R':
            min_col = min_col + ((max_col - min_col + 1) >> 1);
            break;
      }
    }

    found_seats[min_row] |= 1 << min_col;
  }

  for (i = 1; i < 1023; i++) {
    if (    (found_seats[(i - 1) >> 3] & (1 << ((i - 1) & 0x7)))
        && !(found_seats[(i + 0) >> 3] & (1 << ((i + 0) & 0x7)))
        &&  (found_seats[(i + 1) >> 3] & (1 << ((i + 1) & 0x7)))) {
      printf("%d\n", i);
    }
  }

  return 0;
}
